File: backend/src/api/runs.rs
use crate::db::DbPool;
use crate::error::{AppError, AppResult};
use crate::middleware::auth::AuthUser;
use crate::services::run_service::{
    get_all_run_items, get_batch_items, get_batch_summary, get_run_details, list_runs,
    revert_run_status, BatchItemsResponse, BatchSummaryResponse, RunDetailsResponse,
    RunListResponse,
};
use axum::{
    extract::{Path, Query, State},
    Json,
};
use serde::Deserialize;

/// Query parameters for list_runs endpoint
#[derive(Debug, Deserialize)]
pub struct ListRunsQuery {
    #[serde(default = "default_limit")]
    pub limit: i32,

    #[serde(default = "default_offset")]
    pub offset: i32,

    /// Optional search query to filter runs by RunNo, FormulaId, or FormulaDesc
    pub search: Option<String>,
}

fn default_limit() -> i32 {
    10
}

fn default_offset() -> i32 {
    0
}

/// GET /api/runs
///
/// List all production runs with pagination
///
/// # OpenAPI Contract
/// - operationId: listRuns
/// - Query parameters: limit (default 10, max 100), offset (default 0)
/// - Response 200: RunListResponse with runs array and pagination metadata
///
/// # Constitutional Compliance
/// * ✅ JWT authentication required (AuthUser extractor)
/// * ✅ Uses composite keys (RunNo, RowNum) via GROUP BY
/// * ✅ Filters Status IN ('NEW', 'PRINT')
/// * ✅ Pagination with OFFSET/FETCH NEXT (10 per page default)
pub async fn list_runs_endpoint(
    State(pool): State<DbPool>,
    AuthUser(claims): AuthUser,
    Query(params): Query<ListRunsQuery>,
) -> AppResult<Json<RunListResponse>> {
    tracing::info!(
        user = %claims.username,
        limit = params.limit,
        offset = params.offset,
        "GET /api/runs request"
    );

    // Validate query parameters
    if params.limit < 1 {
        return Err(AppError::ValidationError(
            "Limit must be greater than 0".to_string(),
        ));
    }
    if params.limit > 100 {
        return Err(AppError::ValidationError(
            "Limit cannot exceed 100".to_string(),
        ));
    }
    if params.offset < 0 {
        return Err(AppError::ValidationError(
            "Offset cannot be negative".to_string(),
        ));
    }

    let response = list_runs(&pool, params.limit, params.offset, params.search.as_deref()).await?;

    Ok(Json(response))
}

/// GET /api/runs/:runNo
///
/// Get run details with auto-population fields
///
/// # OpenAPI Contract
/// - operationId: getRunDetails
/// - Path parameter: runNo (integer, minimum 1)
/// - Response 200: RunDetailsResponse
/// - Response 404: NotFoundError (Run No not found)
///
/// # Constitutional Compliance
/// * ✅ JWT authentication required (AuthUser extractor)
/// * ✅ Uses composite keys (RunNo, RowNum)
/// * ✅ Auto-populates fgItemKey from FormulaId
/// * ✅ Auto-populates fgDescription from FormulaDesc
pub async fn get_run_details_endpoint(
    State(pool): State<DbPool>,
    AuthUser(claims): AuthUser,
    Path(run_no): Path<i32>,
) -> AppResult<Json<RunDetailsResponse>> {
    tracing::info!(
        user = %claims.username,
        run_no = run_no,
        "GET /api/runs/{} request",
        run_no
    );

    // Validate run_no
    if run_no < 1 {
        return Err(AppError::ValidationError(
            "Run number must be greater than 0".to_string(),
        ));
    }

    let response = get_run_details(&pool, run_no).await?;

    Ok(Json(response))
}

/// GET /api/runs/:runNo/batches/:rowNum/items
///
/// Get items for batch with weight range calculation
///
/// # OpenAPI Contract
/// - operationId: getBatchItems
/// - Path parameters: runNo (integer), rowNum (integer)
/// - Response 200: BatchItemsResponse
/// - Response 404: NotFoundError (Batch not found)
///
/// # Constitutional Compliance
/// * ✅ JWT authentication required (AuthUser extractor)
/// * ✅ Uses composite keys (RunNo, RowNum, LineId)
/// * ✅ Uses PickedPartialQty (NOT PickedPartialQtyKG)
/// * ✅ Weight range = ToPickedPartialQty ± INMAST.User9
pub async fn get_batch_items_endpoint(
    State(pool): State<DbPool>,
    AuthUser(claims): AuthUser,
    Path((run_no, row_num)): Path<(i32, i32)>,
) -> AppResult<Json<BatchItemsResponse>> {
    tracing::info!(
        user = %claims.username,
        run_no = run_no,
        row_num = row_num,
        "GET /api/runs/{}/batches/{}/items request",
        run_no,
        row_num
    );

    // Validate parameters
    if run_no < 1 {
        return Err(AppError::ValidationError(
            "Run number must be greater than 0".to_string(),
        ));
    }
    if row_num < 1 {
        return Err(AppError::ValidationError(
            "Row number must be greater than 0".to_string(),
        ));
    }

    let response = get_batch_items(&pool, run_no, row_num).await?;

    Ok(Json(response))
}

/// GET /api/runs/:runNo/items
///
/// Get ALL items across ALL batches for a run
///
/// Used by ItemSelectionModal to show all unpicked items across all batches
///
/// # OpenAPI Contract
/// - operationId: getAllRunItems
/// - Path parameter: runNo (integer, minimum 1)
/// - Response 200: BatchItemsResponse
/// - Response 404: NotFoundError (Run not found)
///
/// # Constitutional Compliance
/// * ✅ JWT authentication required (AuthUser extractor)
/// * ✅ Uses composite keys (RunNo, RowNum, LineId)
/// * ✅ Returns all items from all batches
pub async fn get_all_run_items_endpoint(
    State(pool): State<DbPool>,
    AuthUser(claims): AuthUser,
    Path(run_no): Path<i32>,
) -> AppResult<Json<BatchItemsResponse>> {
    tracing::info!(
        user = %claims.username,
        run_no = run_no,
        "GET /api/runs/{}/items request",
        run_no
    );

    // Validate run_no
    if run_no < 1 {
        return Err(AppError::ValidationError(
            "Run number must be greater than 0".to_string(),
        ));
    }

    let response = get_all_run_items(&pool, run_no).await?;

    Ok(Json(response))
}

/// GET /api/runs/:runNo/summary
///
/// Get batch summary data for printing
///
/// Used by the PRINT button to generate batch summary labels (4×4" format)
/// Only returns data when run status is 'PRINT' (all items picked)
///
/// # OpenAPI Contract
/// - operationId: getBatchSummary
/// - Path parameter: runNo (integer, minimum 1)
/// - Response 200: BatchSummaryResponse
/// - Response 404: NotFoundError (Run not found or status not PRINT)
///
/// # Constitutional Compliance
/// * ✅ JWT authentication required (AuthUser extractor)
/// * ✅ Uses composite keys (RunNo, RowNum, LineId)
/// * ✅ Only returns picked items (PickedPartialQty > 0)
/// * ✅ Filters by Status = 'PRINT'
/// * ✅ Groups items by batch (RowNum)
pub async fn get_batch_summary_endpoint(
    State(pool): State<DbPool>,
    AuthUser(claims): AuthUser,
    Path(run_no): Path<i32>,
) -> AppResult<Json<BatchSummaryResponse>> {
    tracing::info!(
        user = %claims.username,
        run_no = run_no,
        "GET /api/runs/{}/summary request",
        run_no
    );

    // Validate run_no
    if run_no < 1 {
        return Err(AppError::ValidationError(
            "Run number must be greater than 0".to_string(),
        ));
    }

    let response = get_batch_summary(&pool, run_no).await?;

    Ok(Json(response))
}

/// POST /api/runs/:runNo/revert-status
///
/// Revert run status from PRINT to NEW
///
/// Allows users to make changes after run completion.
/// Only succeeds if current status is PRINT.
///
/// # OpenAPI Contract
/// - operationId: revertRunStatus
/// - Path parameter: runNo (integer, minimum 1)
/// - Response 200: RunDetailsResponse with status='NEW'
/// - Response 400: ValidationError (Status not PRINT)
/// - Response 404: NotFoundError (Run not found)
///
/// # Constitutional Compliance
/// * ✅ JWT authentication required (AuthUser extractor)
/// * ✅ Atomic UPDATE operation
/// * ✅ Preserves audit trail (only changes Status field)
pub async fn revert_run_status_endpoint(
    State(pool): State<DbPool>,
    AuthUser(claims): AuthUser,
    Path(run_no): Path<i32>,
) -> AppResult<Json<RunDetailsResponse>> {
    tracing::info!(
        user = %claims.username,
        run_no = run_no,
        "POST /api/runs/{}/revert-status request",
        run_no
    );

    // Validate run_no
    if run_no < 1 {
        return Err(AppError::ValidationError(
            "Run number must be greater than 0".to_string(),
        ));
    }

    let response = revert_run_status(&pool, run_no).await?;

    Ok(Json(response))
}

File: frontend/src/services/api/runs.ts
/**
 * Runs API Service
 *
 * Service methods for production run endpoints:
 * - getRunDetails: GET /api/runs/{runNo}
 * - getBatchItems: GET /api/runs/{runNo}/batches/{rowNum}/items
 * - completeRun: POST /api/runs/{runNo}/complete
 *
 * Constitutional Requirements:
 * - Contract-first development (matches openapi.yaml)
 * - Last 5 runs cached offline (FIFO eviction)
 * - Network-first for API (fresh data when online)
 * - Cache fallback when offline
 */

import { apiClient } from './client'
import {
  RunDetailsResponse,
  RunListResponse,
  BatchItemsResponse,
  CompleteRunResponse,
  BatchItemDTO,
} from '@/types/api'
import { cacheRun, getCachedRun } from '@/services/cache'

/**
 * List all production runs with pagination and optional search
 *
 * OpenAPI Operation: GET /api/runs
 *
 * No offline cache support - always fetches fresh data from API
 * Used by Run Search Modal for selecting production runs
 *
 * @param limit - Records per page (default 10, max 100)
 * @param offset - Records to skip (default 0)
 * @param search - Optional search query to filter by RunNo, FormulaId, or FormulaDesc
 * @returns Promise<RunListResponse> - Paginated list of runs
 * @throws ErrorResponse - On network error or API error
 */
export async function listRuns(
  limit: number = 10,
  offset: number = 0,
  search?: string
): Promise<RunListResponse> {
  const params: { limit: number; offset: number; search?: string } = { limit, offset }
  if (search && search.trim() !== '') {
    params.search = search.trim()
  }

  const response = await apiClient.get<RunListResponse>('/runs', { params })
  return response.data
}

/**
 * Get run details with offline cache support
 *
 * OpenAPI Operation: GET /api/runs/{runNo}
 *
 * Caching Strategy:
 * 1. Try to fetch from API (Network-first)
 * 2. On success: Cache the run data + batch items
 * 3. On offline: Return cached data if available
 * 4. On error + cached: Return cached data as fallback
 *
 * Constitutional Requirement: Last 5 runs cached (FIFO eviction)
 *
 * @param runNo - Production run number
 * @returns Promise<RunDetailsResponse> - Run details with batches
 * @throws ErrorResponse - On not found (404) or network error (when offline and no cache)
 */
export async function getRunDetails(runNo: number): Promise<RunDetailsResponse> {
  try {
    // Network-first: Try to fetch from API
    const response = await apiClient.get<RunDetailsResponse>(`/runs/${runNo}`)
    const runData = response.data

    // Fetch batch items for all batches to cache complete run data
    const batchItems: BatchItemDTO[] = []
    for (const batchNum of runData.batches) {
      try {
        const items = await getBatchItems(runNo, batchNum)
        batchItems.push(...items)
      } catch (error) {
        console.warn(`[API] Failed to fetch batch ${batchNum} items for caching:`, error)
      }
    }

    // Cache the run data (async - don't wait)
    cacheRun(runNo, runData, batchItems).catch(error => {
      console.error('[API] Failed to cache run data:', error)
    })

    return runData
  } catch (error) {
    // Check if offline (no network connection)
    if (!navigator.onLine) {
      console.warn(`[API] Offline - attempting to use cached data for run ${runNo}`)

      // Try to retrieve cached data
      const cached = await getCachedRun(runNo)
      if (cached) {
        console.log(`[API] Using cached run data for run ${runNo}`)
        return cached.runData
      } else {
        console.error(`[API] No cached data available for run ${runNo}`)
        throw new Error(`Run ${runNo} not available offline (not cached)`)
      }
    }

    // Online but API error - try cache as fallback
    console.warn(`[API] API error - attempting cache fallback for run ${runNo}:`, error)
    const cached = await getCachedRun(runNo)
    if (cached) {
      console.log(`[API] Using cached run data as fallback for run ${runNo}`)
      return cached.runData
    }

    // No cache available - re-throw error
    throw error
  }
}

/**
 * Get items for batch with offline cache support
 *
 * OpenAPI Operation: GET /api/runs/{runNo}/batches/{rowNum}/items
 *
 * Caching Strategy:
 * 1. Try to fetch from API (Network-first)
 * 2. On offline: Return cached batch items if available
 * 3. On error + cached: Return cached data as fallback
 *
 * Note: Batch items are cached together with run details in getRunDetails()
 *
 * @param runNo - Production run number
 * @param rowNum - Batch number (row number)
 * @returns Promise<BatchItemDTO[]> - Batch items with weight ranges
 * @throws ErrorResponse - On not found (404) or network error (when offline and no cache)
 */
export async function getBatchItems(runNo: number, rowNum: number): Promise<BatchItemDTO[]> {
  try {
    // Network-first: Try to fetch from API
    const response = await apiClient.get<BatchItemsResponse>(
      `/runs/${runNo}/batches/${rowNum}/items`
    )
    return response.data.items
  } catch (error) {
    // Check if offline or API error
    if (!navigator.onLine || error) {
      console.warn(`[API] Attempting to use cached batch items for run ${runNo}, batch ${rowNum}`)

      // Try to retrieve cached run data
      const cached = await getCachedRun(runNo)
      if (cached) {
        // Filter cached batch items by rowNum
        // Note: Batch items don't have rowNum field, so we return all cached items
        // The UI should filter by selected batch
        console.log(`[API] Using cached batch items for run ${runNo}`)
        return cached.batchItems
      } else {
        console.error(`[API] No cached data available for run ${runNo}`)
      }
    }

    // No cache available - re-throw error
    throw error
  }
}

/**
 * Complete run and assign pallet
 *
 * OpenAPI Operation: POST /api/runs/{runNo}/complete
 * Execute run completion workflow when all items in all batches are picked:
 * 1. Get next PT sequence number
 * 2. Update run status from NEW to PRINT
 * 3. Create pallet record (Cust_PartialPalletLotPicked)
 * 4. Trigger batch summary label printing
 *
 * @param runNo - Production run number
 * @param workstationId - Workstation identifier (e.g., WS3)
 * @returns Promise<CompleteRunResponse> - Pallet assignment and status
 * @throws ErrorResponse - On validation error (400), not found (404), or not all items picked
 */
export async function completeRun(
  runNo: number,
  workstationId: string
): Promise<CompleteRunResponse> {
  const response = await apiClient.post<CompleteRunResponse>(`/runs/${runNo}/complete`, {
    workstationId,
  })
  return response.data
}

/**
 * Get ALL items across ALL batches for a run
 *
 * OpenAPI Operation: GET /api/runs/{runNo}/items
 *
 * Used by ItemSelectionModal to show all unpicked items across all batches
 *
 * No offline cache support - always fetches fresh data from API
 *
 * @param runNo - Production run number
 * @returns Promise<BatchItemDTO[]> - All items from all batches
 * @throws ErrorResponse - On not found (404) or network error
 */
export async function getAllRunItems(runNo: number): Promise<BatchItemDTO[]> {
  const response = await apiClient.get<BatchItemsResponse>(`/runs/${runNo}/items`)
  return response.data.items
}

/**
 * Runs API Service Object
 * Provides all run-related API methods
 */
export const runsApi = {
  listRuns,
  getRunDetails,
  getBatchItems,
  getAllRunItems,
  completeRun,
}

export default runsApi
